### ISSUE TITLE
Virtualized list issue

### ISSUE THREAD

**Issue by tsunamiq** [snabbdom/snabbdom#1099]
I'm trying to build out a virtualize list within a table and am running into rendering issues.  Following a standard pattern of removing  and adding rows when scrolling through a list on a table. Scroll down, remove nodes from the top and add rows to the bottom while  keeping all indexes up to date with the local state of the entire list. 

issue I'm running into is when I scroll down I'm finding that the entire list re-rerenders after rows are removed from the top of the list.  When scrolling up there is no issue: rows are added to the top of the list and removes from the bottom with existing rows not re-rerendered.

I've narrowed down this issue to this code block: https://github.com/snabbdom/snabbdom/blob/710b319622557196db90422d67a3ed44d48908c6/src/init.ts#L374

deleted rows are pushed down in the dom until they reach the bottom and are removed. This causes all rows to re-render. I believe this is the case. Is it possible to verify and possibly propose a solution to avoid re-rendering all untouched rows when removed top rows/nodes?

**Comment by paldepind** [comment]
Thanks for reporting this and digging into the code to find the cause. I think your diagnosis is correct. It seems that a removed element early in the list together with an added element late in the list is not handled very well.

When you say that the entire list "re-renders" what do you mean? Is this something you observe as degraded performance, with devtools, or something else?

To fix this it would be helpful to have a small test case demonstrating the problem. Would it be possible for you to put together a small example where the problem is evident?

**Comment by jonathanthinh** [comment]
This was observed in chrome devtools. I'm working with a table and I see the <tr>'s rerender in devtools (the html mark up flashes) and on screen (flickers) even when there is no changes to the actual rows. the only change is that the new list removes rows from the top and adds to the bottom. 

There is an interesting work around that solves this issue by breaking up the render into two events. 

1. Remove the two rows at the top of the list and  render 
2. then add rows to the bottom of the list and render. 

removing only the top rows, I believe, forces logic to execute the oldVnodeBottom === newVnodeBottom logic which avoids the re-rendering of existing rows. adding rows to the bottom does not cause issues. Looking into building an example for you.

**Comment by tsunamiq** [comment]
here is a demo of the issue. open debugger and view list. remove items from top and you'll see all rows re-render. add rows to the bottom and only the bottom row updates. [codesandbox](https://codesandbox.io/p/sandbox/vibrant-artem-722w55?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrrg7srh00063b6jb68z8c96%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrrg7srh00023b6j50slvj7v%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrrg7srh00033b6jbiqcupnc%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrrg7srh00053b6j168zdi7y%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B52.181698924687566%252C47.818301075312434%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrrg7srh00023b6j50slvj7v%2522%253A%257B%2522id%2522%253A%2522clrrg7srh00023b6j50slvj7v%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrrgij6700563b6j2s4b3oid%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A11%252C%2522startColumn%2522%253A8%252C%2522endLineNumber%2522%253A11%252C%2522endColumn%2522%253A8%257D%255D%252C%2522filepath%2522%253A%2522%252Findex.html%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clrrgij6700563b6j2s4b3oid%2522%257D%252C%2522clrrg7srh00053b6j168zdi7y%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrrg7srh00043b6j822n1x6g%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clrrg7srh00053b6j168zdi7y%2522%252C%2522activeTabId%2522%253A%2522clrrg7srh00043b6j822n1x6g%2522%257D%252C%2522clrrg7srh00033b6jbiqcupnc%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clrrg7srh00033b6jbiqcupnc%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by paldepind** [comment]
Thanks for the additional information.

> There is an interesting work around that solves this issue by breaking up the render into two events.
> 
>  1. Remove the two rows at the top of the list and  render
> 
>  2. then add rows to the bottom of the list and render.

That makes sense given how I understand the issue.  It's only when removing in the beginning and adding in the end at the exact same time that the case that is not handled correctly occurs.

> here is a demo of the issue. open debugger and view list. remove items from top and you'll see all rows re-render. add rows to the bottom and only the bottom row updates.

I don't think keys are set up correctly in that example? I think `{ attrs: { key: item.key } }` should be `{ key: item.key }`. Additionally when the vnodes that are being patched against are created from `toVNode` they won't have a key set either.

In that example, elements aren't added and removed at the same time, so I don't think it should expose the bug (unless there is a much bigger bug).

Are you interested in submitting a PR with a test case and a fix? Alternatively, I can take a look at this — maybe in the weekend. My initial idea would be to try and maintain symmetry in how the patching is done. Right now adding at the beginning and removing at the end is handled as it should be, so the symmetric thing should be done in the other case.

**Comment by paldepind** [comment]
> Sorry that I am currently a bit absent wrt maintenance, I am currently busy writing a conference paper.

No worries and good luck with your paper 👍 💪

Is it intentional that `childVdom` is reused between the two patches? Because that's not really supported as far as I recall.

**Comment by tsunamiq** [comment]
> Thanks for the additional information.
> 
> > There is an interesting work around that solves this issue by breaking up the render into two events.
> > 
> > 1. Remove the two rows at the top of the list and  render
> > 2. then add rows to the bottom of the list and render.
> 
> That makes sense given how I understand the issue. It's only when removing in the beginning and adding in the end at the exact same time that the case that is not handled correctly occurs.
> 
> > here is a demo of the issue. open debugger and view list. remove items from top and you'll see all rows re-render. add rows to the bottom and only the bottom row updates.
> 
> I don't think keys are set up correctly in that example? I think `{ attrs: { key: item.key } }` should be `{ key: item.key }`. Additionally when the vnodes that are being patched against are created from `toVNode` they won't have a key set either.
> 
> In that example, elements aren't added and removed at the same time, so I don't think it should expose the bug (unless there is a much bigger bug).
> 
> Are you interested in submitting a PR with a test case and a fix? Alternatively, I can take a look at this — maybe in the weekend. My initial idea would be to try and maintain symmetry in how the patching is done. Right now adding at the beginning and removing at the end is handled as it should be, so the symmetric thing should be done in the other case.

ahh yes thank you. I've only used snabbdom for about a week now. I can take a look when I have some more free time this week.

**Comment by jvanbruegge** [comment]
> No worries and good luck with your paper 👍 💪

Thanks :)

> Is it intentional that childVdom is reused between the two patches? Because that's not really supported as far as I recall.

I think so yes. I know that vnode reuse in general is not supported, but I still wanted to fix this as this is very localized (IIRC the change is contained in a single for loop). But yeah, probably this is a different issue then

**Comment by paldepind** [comment]
I see. Thanks for sharing that test case. I think it is a different issue, but good to keep in mind when looking at this one.

**Comment by tsunamiq** [comment]
> Thanks for the additional information.
> 
> > There is an interesting work around that solves this issue by breaking up the render into two events.
> > 
> > 1. Remove the two rows at the top of the list and  render
> > 2. then add rows to the bottom of the list and render.
> 
> That makes sense given how I understand the issue. It's only when removing in the beginning and adding in the end at the exact same time that the case that is not handled correctly occurs.
> 
> > here is a demo of the issue. open debugger and view list. remove items from top and you'll see all rows re-render. add rows to the bottom and only the bottom row updates.
> 
> I don't think keys are set up correctly in that example? I think `{ attrs: { key: item.key } }` should be `{ key: item.key }`. Additionally when the vnodes that are being patched against are created from `toVNode` they won't have a key set either.
> 
> In that example, elements aren't added and removed at the same time, so I don't think it should expose the bug (unless there is a much bigger bug).
> 
> Are you interested in submitting a PR with a test case and a fix? Alternatively, I can take a look at this — maybe in the weekend. My initial idea would be to try and maintain symmetry in how the patching is done. Right now adding at the beginning and removing at the end is handled as it should be, so the symmetric thing should be done in the other case.

Hi @paldepind, for the example above, how would we convert an element to a vnode with keys so the issues can be replicated?  I can look at it before the weekend but if I don't have solution by then it'd be great if you had some time to look into it.

**Comment by tsunamiq** [comment]
@paldepind  heads up. I believe this will fix the issue but I'm currently running into issue running test and committing locally. The fix is to check if the oldStartVnode is found in the new Vnodes. If not, remove it from the dom and increment oldStartIndex. I don't think this will fix the use case when the new Vnodes are not in the same order as the oldVnodes. so it does fix my issue but not out of ordered lists. The later will have the same behavior as the original code. 

```js
 if (oldKeyToIdx === undefined) {
oldKeyToIdx = createKeyToIdx(oldCh, oldStartIdx, oldEndIdx);
}
if (newKeyToIdx === undefined) {
newKeyToIdx = createKeyToIdx(newCh, newStartIdx, newEndIdx);
}

    idxInOld = oldKeyToIdx[newStartVnode.key as string];
    idxInNew = newKeyToIdx[oldStartVnode.key as string];

    // checks if the startVNode is in the new keyToIndex map and if is not found remove from dom.
    if (isUndef(idxInNew)) {
      const ch = oldCh[oldStartIdx];
      api.removeChild(parentElm, ch.elm!);
      oldStartVnode = oldCh[++oldStartIdx];
    } else {
      if (isUndef(idxInOld)) {
        // New element
        api.insertBefore(
          parentElm,
          createElm(newStartVnode, insertedVnodeQueue),
          oldStartVnode.elm!
        );
      } else {
        elmToMove = oldCh[idxInOld];
        if (elmToMove.sel !== newStartVnode.sel) {
          api.insertBefore(
            parentElm,
            createElm(newStartVnode, insertedVnodeQueue),
            oldStartVnode.elm!
          );
        } else {
          patchVnode(elmToMove, newStartVnode, insertedVnodeQueue);
          oldCh[idxInOld] = undefined as any;
          api.insertBefore(parentElm, elmToMove.elm!, oldStartVnode.elm!);
        }
      }
      newStartVnode = newCh[++newStartIdx];
    }
  }

**Comment by paldepind** [comment]
> Hi @paldepind, for the example above, how would we convert an element to a vnode with keys so the issues can be replicated?

The easiest thing would be to reuse the previous vnode tree again instead of using `toVNode` :)

> I believe this will fix the issue but I'm currently running into issue running test and committing locally. 

Let us know if you need help with that. Thanks for sharing the fix, I'll look more into this over the weekend.

**Comment by paldepind** [comment]
Please check if #1106 fixes the problem in your case :)

**Comment by tsunamiq** [comment]
@paldepind  your solution looks great! very clean and I really like the bottom up approach.  It'll definitely work for us. Thank you! you're the 🐐

**Comment by paldepind** [comment]
You're welcome and thanks for the kind words :)

This optimization/fix is included in the new 3.6.2 release.
